src/modules/musig/tests_impl.h

```diff
@@ -125,616 +169,459 @@ void musig_api_tests(secp256k1_scratch_space *scratch) {
-        ecount = 0;
-        CHECK(secp256k1_musig_partial_sig_verify(ctx_tmp, &session[1], &signers1[0], &partial_sig[0], &pk[0]) == 0);
-        CHECK(ecount == 1);
-        CHECK(secp256k1_musig_partial_sign(ctx_tmp, &session[1], &partial_sig[1]) == 0);
-        CHECK(ecount == 2);
-        memset(&partial_sig[1], 0, sizeof(partial_sig[1]));
-        CHECK(secp256k1_musig_partial_sig_combine(ctx_tmp, &session[1], sig, partial_sig, 2) == 0);
-        CHECK(ecount == 3);
-
-        CHECK(secp256k1_musig_session_combine_nonces(ctx, &session[1], signers1, 2, NULL, NULL) == 1);
-        CHECK(secp256k1_musig_partial_sig_verify(ctx, &session[1], &signers1[0], &partial_sig[0], &pk[0]) == 1);
-        /* messagehash should be the same as a session whose get_public_nonce was called
-         * with different signers (i.e. they diff in public keys). This is because the
-         * public keys of the signers is set in stone when initializing the session. */
-        secp256k1_musig_compute_messagehash(ctx, msghash1, &session[1]);
-        musig_state_machine_diff_signer_msghash_test(msghash2, pk, &combined_pk, &pre_session, ncs, msg, nonce[0], sk[1], session_id[1]);
-        CHECK(memcmp(msghash1, msghash2, 32) == 0);
-        CHECK(secp256k1_musig_partial_sign(ctx, &session[1], &partial_sig[1]) == 1);
-
-        CHECK(secp256k1_musig_partial_sig_verify(ctx, &session[1], &signers1[1], &partial_sig[1], &pk[1]) == 1);
-        /* Wrong signature */
-        CHECK(secp256k1_musig_partial_sig_verify(ctx, &session[1], &signers1[1], &partial_sig[0], &pk[1]) == 0);
-        /* Can't get the public nonce until msg is set */
-        musig_state_machine_late_msg_test(pk, &combined_pk, &pre_session, nonce_commitment[0], nonce[0], sk[1], session_id[1], msg);
+    unsigned char agg_pk[32];
+    unsigned char extra_input[32];
+    int i, j;
+    secp256k1_scalar k[5][2];
+
+    secp256k1_rfc6979_hmac_sha256_generate(&secp256k1_test_rng, session_id, sizeof(session_id));
+    secp256k1_rfc6979_hmac_sha256_generate(&secp256k1_test_rng, sk, sizeof(sk));
+    secp256k1_rfc6979_hmac_sha256_generate(&secp256k1_test_rng, msg, sizeof(msg));
+    secp256k1_rfc6979_hmac_sha256_generate(&secp256k1_test_rng, agg_pk, sizeof(agg_pk));
+    secp256k1_rfc6979_hmac_sha256_generate(&secp256k1_test_rng, extra_input, sizeof(extra_input));
+
+    /* Check that a bitflip in an argument results in different nonces. */
+    args[0] = session_id;
+    args[1] = sk;
+    args[2] = msg;
+    args[3] = agg_pk;
+    args[4] = extra_input;
+    for (i = 0; i < count; i++) {
+        musig_nonce_bitflip(args, 0, sizeof(session_id));
+        musig_nonce_bitflip(args, 1, sizeof(sk));
+        musig_nonce_bitflip(args, 2, sizeof(msg));
+        musig_nonce_bitflip(args, 3, sizeof(agg_pk));
+        musig_nonce_bitflip(args, 4, sizeof(extra_input));
+    }
+    /* Check that if any argument is NULL, a different nonce is produced than if
```

Comment:
Did you intend to use musig_nonce_null for these? If not, should musig_nonce_null be deleted (since it's unused)?

Reply:
I probably intended to use nonce_null, then wrote the nested loop. The nested loop is better because it detects a missing marker in the nonce function. I deleted musig_nonce_null